City-of-Helsinki/hel-fi-drupal-grants
# feat: AU-2036: E2E, Refactor form tests

# [AU-2528](https://helsinkisolutionoffice.atlassian.net/browse/AU-2528)

## What was done

This pull request refactors all form tests by extracting all test functions to a separate file, `test_generator_helpers.ts`. Previously, every form test file implemented their own set of tests, which caused a lot of duplication. Now, each test file simply generates their tests by passing in their data to the `generateTests()` function like so:

private_person_48.ts
![Screenshot 2024-06-07 at 9 48 46](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/60c33211-5b15-4ca7-8964-94ebaef55837)


## How to install

Make sure your instance is up and running on correct branch.
* `feature/AU-2036-e2e-refactor-form-tests`
* `make fresh`
* `make drush-cr`

## How to test

- Review the test implementations in the form test files. Make sure everything looks alright. Every implementation should pretty much be identical, expect for the `profileType` and `formId` values of course.

- Review the code in the `test_generator_helpers.ts` file

- Remove `ENABLED_FORM_VARIANTS` and `DISABLED_FORM_VARIANTS` from the `.env` file if you want to run all form variants. If you're in a hurry, you can set `ENABLED_FORM_VARIANTS="draft"`.

- Run a few test to make sure everything still works normally: `make test-pw-p PROJECT=forms-64 forms-48`. I ran every test twice, and everything seemed OK:

![Screenshot 2024-06-06 at 22 59 45](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/cfaa8d47-5c58-42db-8b34-a25ceb02c7f1)

<img width="1364" alt="Screenshot 2024-06-07 at 11 39 07" src="https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/9609e960-5339-4b70-aabb-29158788b96b">



[AU-2528]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2528?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
